pagopa/io-app
# feat: [IOCOM-1413] Archiving/Restoring of user messages, new Messages Home

⚠️ This PR depends on #5912 ⚠️

## Short description
This PR adds the support for archiving and restoring of user messages, in the new messages home.

|iOS|Android|
|-|-|
|<video src="https://github.com/pagopa/io-app/assets/5150343/a5f7a2a9-3c20-4ea6-9f34-ce18f22b7ee9" />|<video src="https://github.com/pagopa/io-app/assets/5150343/025fcc9f-3123-44bf-b45a-b5d4d36cfa4d" />|

**_Please note_** that the toast showing "Annullato" has been removed.

## List of changes proposed in this pull request
New archiving/restoring UI and UX. In order to both fix previous problems and not change a complex business logic, this PR builds above the latter, keeping the compatibility with the old system (the algorithm is explained in `ts/features/messages/saga/handleUpsertMessageStatusAttributes.ts`)

- actions and reducer. The feature is implemented with three states:
  - "disabled" | "enabled": the latter displays the bottom CTAs and allows to select messages for archiving/restoring
  - "processing": the system is busy processing the asynchronous archiving/restoring of messages

Messages are enqueued by Id while in the "enabled" state and processed one at a time while in "processing" state, in order to properly handle whatever failure may happen. It also supports cancellation by the user. If at any moment an error or a cancellation happens, the processing is suspended, the user is notified but she can later resume it.

- `ArchiveRestoreBar` is the component that displays the bottom CTAs for archiving/restoring

Since the bottom bar is hidden while picking messages or processing them, there are some application entry points that have been disabled, in order not to trigger a flow that may later bring back to another main screen without the bottom bar. Such flows are deep link following, push notification tapping, message reloading and next page loading during processing, first message page loading and message search.

## How to test
Using the io-dev-api-server, make sure to enable both DS and new messages home. Long tap (and later tap) on some messages and try the archiving/restoring/cancelling.
